pyfifotax/data_structures_fifo.py

```diff
@@ -0,0 +1,223 @@
+        self.assets: list[FIFOObject] = []
+        self.total_quantity: decimal.Decimal = to_decimal(0)
+        self.is_eur_queue = is_eur_queue
+        if self.is_eur_queue:
+            self.assets.append(
+                FIFOForex("EUR", to_decimal(0), datetime.date(1, 1, 1), "init")
+            )
+
+    def apply_split(self, shares_after_split: decimal.Decimal):
+        if self.is_eur_queue:
+            return
+
+        self.total_quantity = to_decimal(0)
+        for asset in self.assets:
+            asset.quantity = asset.quantity * shares_after_split
+            asset.buy_price = asset.buy_price / shares_after_split
+            self.total_quantity = self.total_quantity + asset.quantity
+
+    def push(self, asset: FIFOObject):
+        if self.is_empty():
+            self.assets = [asset]
+
+        else:
+            if self.is_eur_queue:
+                self.assets[0].quantity += asset.quantity
+            else:
+                # insert based on buy date ("first in")
+                idx = 0
+                while (idx < len(self.assets)) and (
+                    asset.buy_date > self.assets[idx].buy_date
+                ):
+                    idx += 1
+                self.assets.insert(idx, asset)
+
+        self.total_quantity += asset.quantity
+
+    def is_empty(self):
+        return len(self.assets) == 0
+
+    def peek(self):
+        if self.is_empty():
+            raise ValueError("Cannot peek first element from an empty queue.")
+        return self.assets[0]
+
+    def pop(
+        self,
+        quantity: decimal.Decimal,
+        sell_price: decimal.Decimal,
+        sell_date: datetime.date,
+    ):
+        if self.is_eur_queue:
+            pop_asset = from_asset(self.peek(), quantity)
+            # not really relevant, but leave them anyways
+            pop_asset.sell_price = sell_price
+            pop_asset.sell_date = sell_date
+            self.peek().quantity -= quantity
+            self.total_quantity -= quantity
+            return [pop_asset]
+
+        if math.isclose(quantity, 0, abs_tol=1e-10):
+            return []
+
+        if quantity < 0:
+            raise ValueError("Cannot sell negative amount of equity")
+
+        if quantity > 0 and self.is_empty():
+            raise ValueError(f"Cannot sell equities because there isn't any owned")
+
+        if (
+            not math.isclose(quantity, self.total_quantity)
+            and quantity > self.total_quantity
+        ):
+            symbol = self.peek().symbol
+            asset_type = self.peek().__class__.__name__
+            if asset_type == "FIFOShare":
+                raise ValueError(
+                    f"Cannot sell more {symbol} shares ({quantity:.2f}) than owned overall ({self.total_quantity:.2f})."
+                )
+            elif asset_type == "FIFOForex":
+                raise ValueError(
+                    f"Cannot convert more {symbol} ({quantity:.2f}) than owned overall ({self.total_quantity:.2f})."
```

Comment:
I bumped into a rounding problem, however not with EUR but with Forex. If you go back and forth between EUR and another currency, it can make the other currency balance off a bit. This is how I got this error message:

```
[CRITICAL] Cannot convert more USD (3.80) than owned overall (3.73).
```

I don't think there's a proper solution (apart from noting the exact _to_ and _from_ values during conversion) just add a little bit of wiggle room like 1¤.

Thinking about it, it should be applicable only for currencies (FIFOForex) and not for shares.

Reply:
I thought that I tried to avoid this as much as possible by always using "foreign" amount, i.e. when converting EUR->USD and back, the USD amount technically should always be consistent, but I can easily see where slight rounding errors can happen.
But yes, I can see how a little bit of wiggle room can help in that regard.

Reply:
I haven't tried the fix yet, but according to my experience, this can reach around 1¤ (USD, etc.)

Reply:
I think that the symbol behind 1 isn't properly resolved on my end. Which threshold would you prefer? And would you prefer a relative or an absolute wiggle room?

Reply:
That's the correct symbol :).
¤ is the [currency sign](https://en.wikipedia.org/wiki/Currency_sign_(generic)). It's just not used too often.

Absolute 1¤ should be enough for a start, I think. If someone complains, then we can increase it.